solr/solrj/src/java/org/apache/solr/common/util/CompressionUtil.java

```diff
@@ -0,0 +1,109 @@
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.zip.DataFormatException;
+import java.util.zip.Deflater;
+import java.util.zip.Inflater;
+
+/**
+ * Utility class for dealing with LZ4 compressed data, such as state.json in Zookeeper if configured to be compressed
+ */
+public class CompressionUtil {
+
+    private final static char[] ZLIB_FIRST_MAGIC_HEX = new char[]{'7', '8'};
+    private final static char[] ZLIB_SECOND_MAGIC_HEX = new char[]{'0', '1'};
+    private final static int COMPRESSED_SIZE_MAGIC_NUMBER = 8;
+
+    /**
+     * Uses the hex magic number for zlib compression '78 01' to check if the bytes are compressed
+     * @param data - the bytes to check for compression
+     * @return true if the data is zlib compressed based on the first 2 bytes, false otherwise
+     */
+    public static boolean isCompressedBytes(byte[] data) {
+        if (data == null || data.length < 2) return false;
+        // two characters form the hex value.
+        char[] hexDigits = new char[2];
+        hexDigits[0] = Character.forDigit((data[0] >> 4) & 0xF, 16);
+        hexDigits[1] = Character.forDigit((data[0] & 0xF), 16);
+        if (!Arrays.equals(hexDigits, ZLIB_FIRST_MAGIC_HEX)) return false;
+        hexDigits = new char[2];
+        hexDigits[0] = Character.forDigit((data[1] >> 4) & 0xF, 16);
+        hexDigits[1] = Character.forDigit((data[1] & 0xF), 16);
+        return Arrays.equals(hexDigits, ZLIB_SECOND_MAGIC_HEX);
+    }
+
+    /**
+     * Decompresses zlib compressed bytes, returning the uncompressed data as a byte[]
+     * @param data the input zlib compressed data to decompress
+     * @return the decompressed bytes
+     * @throws DataFormatException - The data is not zlib compressed data
+     */
+    public static byte[] decompressBytes(byte[] data) throws DataFormatException {
+        if (data == null) return null;
+        Inflater inflater = new Inflater();
+        inflater.setInput(data, 0, data.length);
+        // Attempt to get the decompressed size from trailing bytes, this will be present if compressed by Solr
+        int decompressedSize = ByteBuffer.wrap(data, data.length - 8, 4).getInt();
+        int xoredSize = ByteBuffer.wrap(data, data.length - 4, 4).getInt();
+        if ((decompressedSize ^ COMPRESSED_SIZE_MAGIC_NUMBER) != xoredSize) {
+            // Take best guess of decompressed size since it wasn't included in trailing bytes
+            decompressedSize = (int)(1.8*data.length);
+        }
+        byte[] buf = new byte[decompressedSize];
+        int actualDecompressedSize = 0;
+        while (!inflater.finished()) {
+            if (actualDecompressedSize >= buf.length) {
+                buf = Arrays.copyOf(buf, (int)(buf.length*1.5));
+            }
+            actualDecompressedSize += inflater.inflate(buf, actualDecompressedSize, decompressedSize);
+        }
+        return Arrays.copyOf(buf, actualDecompressedSize);
```

Comment:
part of the goal of recording decompressed size is to avoid unnecessary arraycopies. In the common case where we know the exact decompressed size, we can still verify it, but we should almost always be able to directly return `buf` here instead of a copy of `buf` ...